### PR TITLE
Restore RDS instance from a MySQL backup

### PR DESCRIPTION
**Pull request by danports**
Adds support for restoring an RDS instance from a MySQL backup stored in an S3 bucket. Basic version of the functionality discussed in #227.

Fixes https://github.com/crossplane/provider-aws/issues/227

I have:

- [x] Read and followed Crossplane's contribution process.
- [x] Written tests
- [x] Run `make reviewable test` to ensure this PR is ready for review.
- [x] Tested on AWS by deploying a small (<1GB) MySQL backup from a bucket
- [x] Tested on AWS by deploying a medium (~35GB) MySQL backup from a bucket

Open to feedback on the API shape...just wanted to get something working for now. This my first time writing Go code, so this code probably needs a bit of work...some of it is not very DRY.

### CODE REVIEW

File: apis/database/v1beta1/rdsinstance_types.go
Comment (chlunde) [inline apis/database/v1beta1/rdsinstance_types.go:394]: I think this name should take into consideration what the name/structure for restoring from DB snapshots should be too. If the structure will be specific to S3, maybe S3 should be part of the name?

`RestoreDBInstanceFromS3`
`RestoreDBInstanceFromDBSnapshot`
`RestoreDBInstanceToPointInTime`

Comment (negz) [reply apis/database/v1beta1/rdsinstance_types.go:394]: (Disclaimer: I'm not familiar with RDS restores apart from a cursory glance at the API).

If there are in fact multiple different sources you can restore from (e.g. snapshot, S3) perhaps using a type distinction here would make sense. This is a pattern we frequently use in our APIs, e.g.:

```yaml
spec:
  forProvider:
    fromBackUp:
      source: S3
      s3:
        s3bucketName: ...
```

I don't love that we're having to do so much interpretation of the APIs. Ideally our Crossplane managed resource would map directly to one API type (e.g. `CreateDBInstance`), while here we're decorating that API type with another. Given how imperative this call is and that it in fact results in the creation of a DB instance (AFAICT) I can't think of any great alternatives though.

File: pkg/controller/database/rdsinstance.go
Comment (negz) [inline pkg/controller/database/rdsinstance.go:176]:
```suggestion
		if _, err = req.Send(ctx); err != nil {
```

Just a little style nit. Feel free to ignore, but the above suggestion is arguably more idiomatic in Go.

File: apis/database/v1beta1/rdsinstance_types.go
Comment (danports) [reply apis/database/v1beta1/rdsinstance_types.go:394]: Thanks, this feedback was exactly what I was looking for - I am not familiar enough with Crossplane's API patterns to understand how you guys would model a subclass/discriminator case like this. I'll update the model. Yeah, I don't really see any other options either given that the restore operation is a separate API call - it would be nicer if we could pass the restore parameters to the `CreateDBInstance` call, but it doesn't support that.

File: pkg/controller/database/rdsinstance.go
Comment (danports) [reply pkg/controller/database/rdsinstance.go:176]: Sure, thanks for the feedback - this was some copypasta from existing code but I'll update it.

**Comment by stephanedicesare** [comment]
We are also interested in this functionality.
Is there a timeline when this can be merged?

**Comment by jbhennin** [comment]
@danports have you considered also adding this functionality to DBCluster and DBInstance?  This would allow for feature parity and then the restore from snapshot and Pitr can be fast followers.

**Comment by haarchri** [comment]
@chlunde  @muvaf  what are the open points here ? so rebase is open but in general the test is working

**Comment by chlunde** [comment]
@haarchri I think this is the latest comment: https://github.com/crossplane/provider-aws/pull/770#discussion_r685420404 - so maybe the API should be written as negz shows to make it easier to implement the other types of restores (and easier to use).

I think the next step should be to first list all the unique parameters for each restore type, and then decide here in this issue if it makes sense to have the union of all restore options in the same struct (which is the approach the PR currently leads to) or one struct/type for each type of restore.

**Comment by jbhennin** [comment]
@chlunde @haarchri @muvaf Just curious on how this PR stands.  We are really interested in the ability to restore from a snapshot.

**Comment by danports** [comment]
This PR has been on the back burner for me since I needed the feature for a one-time migration that has long since been completed - I don't really have a use case for it anymore. But I'd be happy to address the feedback I received earlier and then leave it to others to take it a step further by adding support for other restore types, etc.

**Comment by danports** [comment]
> I think the next step should be to first list all the unique parameters for each restore type, and then decide here in this issue if it makes sense to have the union of all restore options in the same struct (which is the approach the PR currently leads to) or one struct/type for each type of restore.

@chlunde I took a brief look at the parameters for the other restore types (DB snapshot, point in time) and there isn't really any overlap between the parameters for the different restore types, so the type distinction structure that @negz suggested earlier would probably make the most sense here. I will push a new version of the PR with these changes soon.

File: pkg/controller/database/rdsinstance.go
Comment (danports) [reply pkg/controller/database/rdsinstance.go]: Resolved by prior changes to the code due to AWS SDK updates.

File: apis/database/v1beta1/rdsinstance_types.go
Comment (danports) [reply apis/database/v1beta1/rdsinstance_types.go]: Introduced type distinction in the latest version of the PR.

**Comment by danports** [comment]
I need to test the updated code on AWS, but otherwise it should be good to go...please let me know if you have more feedback @chlunde @negz. Happy 2022! 🥂

**Comment by danports** [comment]
OK, lint issues should be gone now. Had a problem running `make lint` on my machine earlier, but was able to [fix](https://github.com/golangci/golangci-lint/issues/1219#issuecomment-656793602) that today.

**Comment by haarchri** [comment]
@chlunde can you check ?

**Comment by danports** [comment]
Tested my updated code on AWS and everything is looking good! 🥳 Let me know if there are any last nits to address here @chlunde.

**Comment by haarchri** [comment]
I tested as well without issues LGTM @chlunde please do final review

File: apis/database/v1beta1/rdsinstance_types.go
Comment (chlunde) [inline apis/database/v1beta1/rdsinstance_types.go:112]: ideally, this and below would be referencers, but personally I don't mind this as a followup-PR at this point, since you've had to wait so long.

Here's an example: https://github.com/crossplane/provider-aws/commit/c9a8b38ae93a01cff5dd1ca7367f226d1eafecda
Docs: https://github.com/crossplane/crossplane-tools/pull/35

File: pkg/clients/rds/rds.go
Comment (chlunde) [inline pkg/clients/rds/rds.go:137]: maybe leave a short comment/note inside `GenerateCreateDBInstanceInput` that any changes there should be applied in `GenerateRestoreDBInstanceInput`. It makes it more likely to pick up a bug either during implementation or code review.

**Review by chlunde — CHANGES_REQUESTED** [review]
Looks very good, thanks for your contribution and patience! Due to the necessary duplication I think a friendly note for the future reviewers would be nice, except for that it looks good to me.

File: apis/database/v1beta1/rdsinstance_types.go
Comment (muvaf) [inline apis/database/v1beta1/rdsinstance_types.go:404]:
```suggestion
	RestoreFrom *RestoreBackupConfiguration `json:"restoreFrom,omitempty"`
```

The optional fields should [have](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) `omitempty` tag.

File: apis/database/v1beta1/rdsinstance_types.go
Comment (muvaf) [inline apis/database/v1beta1/rdsinstance_types.go:119]:
```suggestion
	Prefix *string `json:"prefix,omitempty"`
```

**Review by muvaf — APPROVED** [review]
Sorry this was assigned to me for review and has been open for so long and thanks for your patience. It looks good to me in general with a few small notes. @chlunde feel free to merge it once your concerns are addressed.

**Comment by muvaf** [comment]
Also feel free to implement it for `DBInstance` in `rds` group as well, which would be mostly copy-paste, in another PR.

File: pkg/clients/rds/rds.go
Comment (danports) [reply pkg/clients/rds/rds.go:137]: Added a comment to both functions.

File: apis/database/v1beta1/rdsinstance_types.go
Comment (danports) [reply apis/database/v1beta1/rdsinstance_types.go:112]: Yeah, let's leave that for a later time - I would like to get this out the door without changes that would require additional testing, etc. Referencers look pretty neat though!

**Review by haarchri — APPROVED** [review]
all concerns are addressed - did a final test - everything is working as expected

thanks @danports for implementation

**Review by chlunde — APPROVED** [review]
Looks good to me, thank you for your contribution.

**Pull request merged**
